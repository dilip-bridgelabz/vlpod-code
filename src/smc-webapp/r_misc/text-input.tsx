/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

import * as React from "react";
import { Icon } from "./icon";
const { Button, FormControl, FormGroup } = require("react-bootstrap");

interface Props {
  text: string;
  on_change: (value: string) => void;
  type?: string;
  rows?: number;
  autoFocus?: boolean;
}

interface State {
  text: string;
}

export class TextInput extends React.Component<Props, State> {
  constructor(props) {
    super(props);
    this.state = { text: props.text };
  }

  componentWillReceiveProps(next_props) {
    if (this.props.text !== next_props.text) {
      // so when the props change the state stays in sync (e.g., so save button doesn't appear, etc.)
      this.setState({ text: next_props.text });
    }
  }

  saveChange = (event) => {
    event.preventDefault();
    this.props.on_change(this.state.text);
  };

  render_save_button() {
    if (this.state.text != undefined && this.state.text !== this.props.text) {
      return (
        <Button
          style={{ marginBottom: "15px" }}
          bsStyle="success"
          onClick={this.saveChange}
        >
          <Icon name="save" /> Save
        </Button>
      );
    }
  }

  render_input() {
    return (
      <FormGroup>
        <FormControl
          type={this.props.type != undefined ? this.props.type : "text"}
          ref="input"
          rows={this.props.rows}
          componentClass={this.props.type === "textarea" ? "textarea" : "input"}
          value={
            this.state.text != undefined ? this.state.text : this.props.text
          }
          onChange={(e) => this.setState({ text: e.target.value })}
          autoFocus={this.props.autoFocus}
        />
      </FormGroup>
    );
  }

  render() {
    return (
      <form onSubmit={this.saveChange}>
        {this.render_input()}
        {this.render_save_button()}
      </form>
    );
  }
}
