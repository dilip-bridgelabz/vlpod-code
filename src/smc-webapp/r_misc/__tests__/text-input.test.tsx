/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

import * as React from "react";
import { shallow } from "enzyme";
import { TextInput } from "../text-input";

test("smoke test", () => {
  const mock_change = jest.fn();
  const starting_text = "initial";
  const changed_text = "changed!";
  const updated_text = "changed from props";

  const rendered = shallow(
    <TextInput text={starting_text} on_change={mock_change} />
  );
  expect(rendered).toMatchSnapshot("initial render");

  const text_input = rendered.find("FormControl").first();
  text_input.simulate("change", { target: { value: changed_text } });
  expect(rendered).toMatchSnapshot(
    "save button appears after text changes from initial render"
  );

  const send_button = rendered.find("Button").first();
  send_button.simulate("click", new MouseEvent("click"));
  expect(mock_change.mock.calls.length).toBe(1);
  expect(mock_change.mock.calls[0][0]).toBe(changed_text);

  rendered.setProps({ text: updated_text });
  expect(rendered).toMatchSnapshot("value updated from props");
});
