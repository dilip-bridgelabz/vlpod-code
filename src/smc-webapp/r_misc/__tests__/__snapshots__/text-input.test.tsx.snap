// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke test: initial render 1`] = `
<form
  onSubmit={[Function]}
>
  <FormGroup
    bsClass="form-group"
  >
    <FormControl
      bsClass="form-control"
      componentClass="input"
      onChange={[Function]}
      type="text"
      value="initial"
    />
  </FormGroup>
</form>
`;

exports[`smoke test: save button appears after text changes from initial render 1`] = `
<form
  onSubmit={[Function]}
>
  <FormGroup
    bsClass="form-group"
  >
    <FormControl
      bsClass="form-control"
      componentClass="input"
      onChange={[Function]}
      type="text"
      value="changed!"
    />
  </FormGroup>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="success"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "marginBottom": "15px",
      }
    }
  >
    <Icon
      name="save"
    />
     Save
  </Button>
</form>
`;

exports[`smoke test: value updated from props 1`] = `
<form
  onSubmit={[Function]}
>
  <FormGroup
    bsClass="form-group"
  >
    <FormControl
      bsClass="form-control"
      componentClass="input"
      onChange={[Function]}
      type="text"
      value="changed from props"
    />
  </FormGroup>
</form>
`;
